pikelet/src/surface/lex.rs

```diff
@@ -0,0 +1,190 @@
+use logos::Logos;
+use std::fmt::{self, Display, Formatter};
+
+/// The complete set of `LexToken`s some of which never escape the lexer.
+/// See Token for a list of which Tokens do and do not escape.
+#[derive(Logos)]
+enum LexToken {
+    #[end]
+    EOF,
+    #[token = ":"]
+    Colon,
+    #[token = ","]
+    Comma,
+    #[token = "fun"]
+    Fun,
+    #[token = "=>"]
+    DArrow,
+    #[token = "->"]
+    Arrow,
+    #[token = "("]
+    LParen,
+    #[token = ")"]
+    RParen,
+    #[token = "["]
+    LBrack,
+    #[token = "]"]
+    RBrack,
+    #[token = "{"]
+    LBrace,
+    #[token = "}"]
+    RBrace,
+    #[token = r"record"]
+    RecordTerm,
+    #[token = r"Record"]
+    RecordType,
+    #[token = "."]
+    Dot,
+    #[token = "="]
+    Equal,
+    // Hmm, not sure why this doesn't work.
+    // #[regex = r#"'(.|\\"|\\')*'"#]
+    #[regex = r#"('[^'\\]|\\t|\\u|\\n|\\"|\\')*'"#]
+    CharLit,
+    // Ditto.
+    // #[regex = r#""(.|\\"|\\')*""#]
+    #[regex = r#""([^"\\]|\\t|\\u|\\n|\\"|\\')*""#]
+    StrLit,
+    #[regex = r"[-+]?[0-9]+(\.[0-9]+)?"]
+    Number,
+    #[regex = r"[a-zA-Z][a-zA-Z0-9\-]*"]
+    Name,
+    #[regex = r"\^[0-9]+(\.[0-9]+)?"]
+    Shift,
+    #[regex = r"\p{Whitespace}"]
+    Whitespace,
+    #[error]
+    Error,
+}
+
+/// The subset of `LexToken`s for the parser.
+/// The tokens in `LexToken`s which are excluded from this enum are:
+/// * Whitespace -- skipped.
+/// * EOF -- turned into None, rather than a token.
+/// * Error -- turned into Some(Err(...)).
+#[derive(Debug, Clone)]
+pub enum Token<'a> {
+    Colon,
+    Comma,
+    Fun,
+    DArrow,
+    Arrow,
+    LParen,
+    RParen,
+    LBrack,
+    RBrack,
+    LBrace,
+    RBrace,
+    RecordTerm,
+    RecordType,
+    Equal,
+    Dot,
+    CharLit(&'a str),
+    StrLit(&'a str),
+    Number(&'a str),
+    Name(&'a str),
+    Shift(&'a str),
+}
+
+impl<'a> Display for Token<'a> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        use Token as T;
```

Comment:
Do we need a local import for this? Might be easier just to match on `Token` on each of the variant.

Reply:
Nope not really.